File: src/sigmatism/clj/sigmatism.clj
(ns sigmatism.clj.sigmatism
  (:refer-clojure :exclude [eval atom assoc not]))

(defn eq [x y]
  (cond (and (= x '()) (= y '()))
        't
        (and (symbol? x) (symbol? y) (= x y))
        't
        :else
        '()))

(defn assoc [x y]
  (if (= 't (eq (ffirst y) x))
    (-> y first fnext)
    (assoc x (rest y))))

(declare eval)

(defn evcon [c a]
  (if (= 't (eval (ffirst c) a))
    (eval (-> c first fnext) a)
    (evcon (rest c) a)))

(defn null [x]
  (eq x '()))

(defn append [x y]
  (if (= 't (null x))
    y
    (cons (first x) (append (rest x) y))))

(defn not [x]
  (if (= 't x)
    '()
    't))

(defn atom [x]
  (if (or (symbol? x)
          (= '() x))
    't
    '()))

(defn pair [x y]
  (cond (and (= 't (null x)) (= 't (null y)))
        '()

        (and (= 't (not (atom x))) (= 't (not (atom y))))
        (cons (list (first x) (first y))
              (pair (rest x) (rest y)))))

(defn evlis [m a]
  (if (= 't (null m))
    '()
    (cons (eval (first m) a)
          (evlis (rest m) a))))

(defn eval [e a]
  (let [[fun & args] e]
    (cond (= 't (atom e))
          (assoc e a)

          (= 't (atom fun))
          (cond
           (= fun 'quote) (first args)
           (= fun 'atom)  (atom (eval (first args) a))
           (= fun 'eq)    (eq (eval (first args) a)
                              (eval (second args) a))
           (= fun 'car)   (first (eval (first args) a))
           (= fun 'cdr)   (rest (eval (first args) a))
           (= fun 'cons)  (cons (eval (first args) a)
                                (eval (second args) a))
           (= fun 'cond)  (evcon args a)
           :else
           (eval (cons (assoc fun a) args) a))

          (= (first fun) 'label)
          (let [[name body] (rest fun)]
            (eval (cons body args)
                  (cons (list name fun) a)))

          (= (first fun) 'lambda)
          (let [[largs body] fun]
            (eval body (append (pair largs (evlis args a)) a))))))

File: src/sigmatism/pl/sigmatism.pl
char(C, [C|A], A) :-
    char_type(C, alnum); char_type(C, period); char_type(C, quote).

chars([C]) --> char(C).
chars([C|Cs]) --> char(C), chars(Cs).

symbol(Symbol, A, R) :-
    append(S, R, A),
    phrase(chars(Cs), S),
    atom_chars(Symbol, Cs).

rest([C]) --> cons(C).
rest([S]) --> symbol(S).
rest([C | R]) --> cons(C), [' '], rest(R).
rest([S1 | S2]) --> symbol(S1), [' '] , rest(S2).

cons([]) --> ['(',')'].
cons(Elems) --> ['('], rest(Elems), [')'].

read_lisp(String, AST) :-
    phrase((cons(AST) ; symbol(AST)), String).

evcon([[F, S] | _], Ns, Value) :-
    eval_lisp(F, Ns, B),
    B \= nil,
    eval_lisp(S, Ns, Value).
evcon([_ | R], Ns, Value) :-
    evcon(R, Ns, Value).

evlis([], _, []).
evlis([Arg | Args], Ns, [Value | R]) :-
    eval_lisp(Arg, Ns, Value),
    evlis(Args, Ns, R).

pair([], _, Ns, Ns).
pair([Key | Keys], [Arg | Args], Ns, Namespace) :-
    put_assoc(Key, Ns, Arg, Assoc),
    pair(Keys, Args, Assoc, Namespace).

eval_lisp(AST, Ns, Value) :-
    atom(AST),
    get_assoc(AST, Ns, Value).
eval_lisp([quote, Q], _, Q).
eval_lisp([atom, A], Ns, t) :-
    eval_lisp(A, Ns, Value),
    atom(Value).
eval_lisp([atom | _], _, nil).
eval_lisp([eq, F, S], Ns, t) :-
    eval_lisp(F, Ns, VF),
    eval_lisp(S, Ns, VS),
    atom(VF),
    atom(VS),
    VF = VS.
eval_lisp([eq | _], _, nil).
eval_lisp([car, F], Ns, Value) :-
    eval_lisp(F, Ns, [Value | _]).
eval_lisp([cdr, F], Ns, Value) :-
    eval_lisp(F, Ns, [_ | Value]).
eval_lisp([cons, F, S], Ns, [V1 | V2]) :-
    eval_lisp(F, Ns, V1),
    eval_lisp(S, Ns, V2).
eval_lisp([cond | R], Ns, Value) :-
    evcon(R, Ns, Value).
eval_lisp([[label, Name, Lambda] | Args], Ns, Value) :-
    put_assoc(Name, Ns, [label, Name | Lambda], Assoc),
    eval_lisp([Lambda | Args], Assoc, Value).
eval_lisp([[lambda, LArgs, Body] | Rest], Ns, Value) :-
    evlis(Rest, Ns, Args),
    pair(LArgs, Args, Ns, Namespace),
    eval_lisp(Body, Namespace, Value).
eval_lisp([H | T], Ns, Value) :-
    get_assoc(H, Ns, E),
    eval_lisp([E | T], Ns, Value).

insert_space([H], [H]).
insert_space([H | T], [H, ' ' | R]) :-
    insert_space(T, R).

writeln_lisp(L) :-
    write_lisp(L), nl.
write_lisp(A) :-
    atom(A),
    write(A).
write_lisp(L) :-
    write('('),
    insert_space(L, P),
    maplist(write_lisp, P),
    write(')').

repl :-
    prompt1('lambda> '),
    current_input(Stream),
    read_line_to_codes(Stream, String),
    maplist(char_code, Chars, String),
    read_lisp(Chars, AST),
    list_to_assoc([], Namespace),
    eval_lisp(AST, Namespace, V),
    writeln_lisp(V),
    repl.
